pty/pty_windows.go

```diff
@@ -134,39 +137,68 @@ func (p *ptyWindows) Resize(height uint16, width uint16) error {
 		Y: int16(height),
 		X: int16(width),
 	})))))
-	if ret != 0 {
+	if windows.Handle(ret) != windows.S_OK {
 		return err
 	}
 	return nil
 }
 
-func (p *ptyWindows) Close() error {
-	p.closeMutex.Lock()
-	defer p.closeMutex.Unlock()
-	if p.closed {
-		return nil
-	}
-	p.closed = true
-
+// closeConsoleNoLock closes the console handle, and sets it to
+// windows.InvalidHandle. It must be called with p.closeMutex held.
+func (p *ptyWindows) closeConsoleNoLock() error {
 	// if we are running a command in the PTY, the corresponding *windowsProcess
 	// may have already closed the PseudoConsole when the command exited, so that
 	// output reads can get to EOF.  In that case, we don't need to close it
 	// again here.
 	if p.console != windows.InvalidHandle {
+		// Because ClosePseudoConsole has no return value, we could ignore the
+		// error here, but we limit this either S_OK or S_FALSE as a safety-net
+		// in case other values could be returned.
+		// https://docs.microsoft.com/en-us/windows/console/closepseudoconsole
 		ret, _, err := procClosePseudoConsole.Call(uintptr(p.console))
-		if ret < 0 {
-			return xerrors.Errorf("close pseudo console: %w", err)
+		if ret := windows.Handle(ret); ret != windows.S_OK && ret != windows.S_FALSE {
+			return xerrors.Errorf("close pseudo console (%d): %w", ret, err)
 		}
 		p.console = windows.InvalidHandle
 	}
 
-	// We always have these files
-	_ = p.outputRead.Close()
-	_ = p.inputWrite.Close()
-	// These get closed & unset if we Start() a new process.
+	return nil
+}
+
+func (p *ptyWindows) Close() error {
+	p.closeMutex.Lock()
+	defer p.closeMutex.Unlock()
+	if p.closed {
+		return nil
+	}
+
+	err := p.closeConsoleNoLock()
+	if err != nil {
+		return err
+	}
+
+	// Only set closed after the console has been successfully closed.
+	p.closed = true
+
+	// Tear down the pipes to unblock readers/writers. Note that outputWrite and
+	// inputRead are unset when we Start() a new process.
+	//
+	// The console is closed so there will be no more writes, we need to do this
+	// here to ensure the discard below is unblocked.
 	if p.outputWrite != nil {
 		_ = p.outputWrite.Close()
 	}
+	// Unless we drain outputRead here, calling Close may wait indefinitely:
+	// https://learn.microsoft.com/en-us/windows/console/closepseudoconsole
+	//
+	// Note that this may result in a small amount of unprocessed output
+	// being lost, but the alternative is to wrap the reader or to handle
+	// this by all consumers.
```

Comment:
> Chiefly, its the [ClosePseudoConsole](https://learn.microsoft.com/en-us/windows/console/closepseudoconsole) call that might block if the output is not drained. By the time you get here, ClosePseudoConsole has already returned.

I would say the docs are unclear. They mention a capital `Close` call in mid-sentence, this could refer to either the `hOutput` buffer or `ClosePseudoConsole` itself (why not just write it out fully if that's the case?). It's very possible you're right, though, it would make more sense.

> Secondly, the [failed test](https://github.com/coder/coder/actions/runs/5442614267/jobs/9898091930?pr=8280) shows that there is a goroutine already attempting to read from the pipe, and getting blocked. Adding an additional reader does not seem like it will help things!

I must've missed the read was active, thanks for pointing it out.